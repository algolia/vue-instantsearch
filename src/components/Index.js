import { createSuitMixin } from '../mixins/suit';
import { createWidgetMixin } from '../mixins/widget';
import indexWidget from 'instantsearch.js/es/widgets/index/index';
import { isVue3, h } from '../util/vue';

// wrapped in a dummy function, since indexWidget doesn't render
const connectIndex = () => indexWidget;

export default {
  name: 'AisIndex',
  mixins: [
    createSuitMixin({ name: 'Index' }),
    createWidgetMixin({ connector: connectIndex }),
  ],
  provide() {
    return {
      // The widget is created & registered by widgetMixin, accessor is needed
      // because provide is not reactive.
      $_ais_getParentIndex: () => this.widget,
    };
  },
  props: {
    indexName: {
      type: String,
      required: true,
    },
    indexId: {
      type: String,
      required: false,
    },
  },
  render(createElement) {
    return (isVue3 ? h : createElement)(
      'div',
      {},
      isVue3 ? this.$slots.default() : this.$slots.default
    );
  },
  computed: {
    widgetParams() {
      return {
        indexName: this.indexName,
        indexId: this.indexId,
      };
    },
  },
};
