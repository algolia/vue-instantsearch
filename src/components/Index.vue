<template>
  <div :class="bem()">
    <slot></slot>
  </div>
</template>

<script>
import { createFromAlgoliaCredentials } from '../store';
import algoliaComponent from '../component';

export default {
  mixins: [algoliaComponent],
  props: {
    searchStore: {
      type: Object,
      default() {
        return this._searchStore;
      },
    },
    apiKey: {
      type: String,
      default() {
        if (this._searchStore) {
          return this._searchStore.algoliaApiKey;
        }

        return undefined;
      },
    },
    appId: {
      type: String,
      default() {
        if (this._searchStore) {
          return this._searchStore.algoliaAppId;
        }

        return undefined;
      },
    },
    indexName: {
      type: String,
      default() {
        if (this._searchStore) {
          return this._searchStore.indexName;
        }

        return undefined;
      },
    },
    query: {
      type: String,
      default: '',
    },
    queryParameters: {
      type: Object,
    },
    cache: {
      type: Boolean,
      default: true,
    },
    autoSearch: {
      type: Boolean,
      default: true,
    },
  },
  data() {
    return {
      blockClassName: 'ais-index',
    };
  },
  provide() {
    if (!this.searchStore) {
      this._localSearchStore = createFromAlgoliaCredentials(
        this.appId,
        this.apiKey
      );
    } else {
      this._localSearchStore = this.searchStore;
    }

    if (this.indexName) {
      this._localSearchStore.indexName = this.indexName;
    }

    if (this.query) {
      this._localSearchStore.query = this.query;
    }

    if (this.queryParameters) {
      this._localSearchStore.queryParameters = this.queryParameters;
    }

    if (this.cache) {
      this._localSearchStore.enableCache();
    } else {
      this._localSearchStore.disableCache();
    }

    return {
      _searchStore: this._localSearchStore,
    };
  },
  mounted() {
    this._localSearchStore.start();
    if (this.autoSearch) {
      this._localSearchStore.refresh();
    }
  },
  watch: {
    indexName() {
      this._localSearchStore.indexName = this.indexName;
    },
    query() {
      this._localSearchStore.query = this.query;
    },
    queryParameters() {
      this._localSearchStore.queryParameters = this.queryParameters;
    },
  },
};
</script>
