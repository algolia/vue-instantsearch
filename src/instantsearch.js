export { createSuitMixin } from './mixins/suit';
export { createWidgetMixin } from './mixins/widget';
export * from './widgets';
export { plugin as default } from './plugin';
export { createInstantSearch } from './util/createInstantSearch';
