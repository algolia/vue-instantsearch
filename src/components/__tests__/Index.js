import { mount } from '@vue/test-utils';
import instantsearch from 'instantsearch.js/es';
import Index from '../Index.vue';

jest.mock('instantsearch.js/es', () => {
  const isPlainObject = require('lodash/isPlainObject');
  const start = jest.fn();
  const fakeInstantSearch = jest.fn(
    ({
      indexName,
      searchClient,
      routing, // eslint-disable-line no-unused-vars
      stalledSearchDelay, // eslint-disable-line no-unused-vars
      searchFunction, // eslint-disable-line no-unused-vars
    }) => {
      if (!searchClient && isPlainObject(searchClient)) {
        throw new Error('need searchClient to be a fn');
      }
      if (!indexName) {
        throw new Error('need indexName to be a string');
      }
      return { start };
    }
  );
  fakeInstantSearch.__startMock = start;
  return fakeInstantSearch;
});

it('passes props to InstantSearch.js', () => {
  const searchClient = {};
  const searchFunction = helper => helper.search();
  mount(Index, {
    propsData: {
      searchClient,
      indexName: 'something',
      routing: true,
      stalledSearchDelay: 250,
      searchFunction,
    },
  });

  expect(instantsearch).toHaveBeenCalledWith({
    indexName: 'something',
    routing: true,
    searchClient,
    searchFunction,
    stalledSearchDelay: 250,
  });
});

it('calls `start` on the next tick', () => {
  mount(Index, {
    propsData: {
      searchClient: {},
      indexName: 'something',
    },
  });

  expect(instantsearch.__startMock).toHaveBeenCalledTimes(1);
});

it('provides an InstantSearch instance', () => {
  const wrapper = mount(Index, {
    propsData: {
      searchClient: {},
      indexName: 'bla',
    },
  });

  expect(wrapper.vm._provided).toEqual({
    instantSearchInstance: {
      // it's really InstantSearch, since it has the same spy as our custom mock
      start: instantsearch.__startMock,
    },
  });
});

it('renders correctly (empty)', () => {
  const wrapper = mount(Index, {
    propsData: {
      searchClient: {},
      indexName: 'bla',
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly (with slot used)', () => {
  const wrapper = mount(Index, {
    propsData: {
      searchClient: {},
      indexName: 'bla',
    },
    slots: {
      default: '<div>Hi there, this is the main slot</div>',
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});
