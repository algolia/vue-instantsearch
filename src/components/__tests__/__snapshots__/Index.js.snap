// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly (empty) 1`] = `

<div class="ais-Index">
</div>

`;

exports[`renders correctly (with slot used) 1`] = `

<div class="ais-Index">
  <div>
    Hi there, this is the main slot
  </div>
</div>

`;
